src/dfa/dfa.rs

```diff
@@ -0,0 +1,601 @@
+use crate::nfa::nfa::NFA;
+use std::collections::{HashMap, HashSet};
+use std::hash::Hash;
+use std::process::id;
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+struct State(String);
+
+enum Tag {
+    Start(usize),
+    End(usize),
+}
+
+struct Transition {
+    from_state: State,
+    symbol: char,
+    to_state: State,
+    tag: Option<Tag>,
+}
+
+pub(crate) struct DFA {
+    start: State,
+    accept: Vec<State>,
+    states: HashSet<State>,
+    transitions: HashMap<State, HashMap<char, Transition>>, // from_state -> symbol -> to_state
+    dfa_to_accepted_nfa_state_mapping: Option<HashMap<State, Vec<(usize, crate::nfa::nfa::State)>>>, // to determine which NFA gets matched
+}
+
+impl DFA {
+    fn new(start_state: State, accept_states: Vec<State>) -> Self {
+        let mut _states = HashSet::new();
+        _states.insert(start_state.clone());
+        for state in accept_states.iter() {
+            _states.insert(state.clone());
+        }
+
+        DFA {
+            start: start_state,
+            accept: accept_states,
+            states: _states,
+            transitions: HashMap::new(),
+            dfa_to_accepted_nfa_state_mapping: None,
+        }
+    }
+
+    fn add_transition(
+        &mut self,
+        from_state: State,
+        symbol: char,
+        to_state: State,
+        tag: Option<Tag>,
+    ) {
+        self.states.insert(from_state.clone());
+        self.states.insert(to_state.clone());
+        self.transitions
+            .entry(from_state.clone())
+            .or_insert_with(HashMap::new)
+            .insert(
+                symbol,
+                Transition {
+                    from_state,
+                    symbol,
+                    to_state,
+                    tag,
+                },
+            );
+    }
+
+    fn simulate(&self, input: &str) -> (Option<HashSet<usize>>, bool) {
```

Comment:
We need a different API that takes char by char since the lexer emits chars as tokens, we don't know the size of string in advance.
In this way, we might also need an API to reset the current simulation state.